File: dataforge-core/src/main/java/hep/dataforge/io/text/HTMLMarkupRenderer.java
package hep.dataforge.io.text;

import hep.dataforge.meta.MetaUtils;
import org.slf4j.LoggerFactory;
import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.Node;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import javax.xml.transform.OutputKeys;
import javax.xml.transform.Transformer;
import javax.xml.transform.TransformerException;
import javax.xml.transform.TransformerFactory;
import javax.xml.transform.dom.DOMSource;
import javax.xml.transform.stream.StreamResult;
import java.io.PrintStream;
import java.util.ArrayDeque;
import java.util.Deque;

/**
 * Created by darksnake on 07-Jan-17.
 */
public class HTMLMarkupRenderer extends GenericMarkupRenderer {
    private Document document;
    private Deque<Element> stack = new ArrayDeque<>();
    private final PrintStream stream;

    public HTMLMarkupRenderer(PrintStream stream) {
        this.stream = stream;
    }

    private Document buildDocument() throws ParserConfigurationException {
        DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
        DocumentBuilder builder = factory.newDocumentBuilder();
        return builder.newDocument();
    }

    private void printDocument() throws TransformerException {
        TransformerFactory tf = TransformerFactory.newInstance();
        Transformer transformer = tf.newTransformer();
        transformer.setOutputProperty(OutputKeys.OMIT_XML_DECLARATION, "yes");
        transformer.setOutputProperty(OutputKeys.INDENT, "yes");
        transformer.setOutputProperty("{http://xml.apache.org/xslt}indent-amount", "4");
        transformer.setOutputProperty(OutputKeys.ENCODING, "utf8");
        transformer.transform(new DOMSource(document), new StreamResult(stream));
        stream.flush();
    }

    @Override
    public void render(Markup element) {
        try {
            document = buildDocument();
            stack.clear();
            Element root = document.createElement("body");
            document.appendChild(root);
            stack.add(root);
            doRender(element);
            stack.removeLast();
            if (stack.size() != 0) {
                LoggerFactory.getLogger(getClass()).warn("Node stack not empty after rendering");
            }
            printDocument();
        } catch (Exception ex) {
            throw new RuntimeException(ex);
        }
    }

    @Override
    protected void renderText(String text, String color, Markup element) {
        if (!color.isEmpty()) {
            Element textElement = document.createElement("font");
            applyHTMLStyle(textElement, element);
            textElement.setAttribute("color", color);
            textElement.setTextContent(text);
            stack.getLast().appendChild(textElement);
        } else {
            Node textElement = document.createTextNode(text);
            stack.getLast().appendChild(textElement);
        }
    }

    @Override
    protected void list(Markup element) {
        Element listNode = document.createElement("ul");
        applyHTMLStyle(listNode, element);
        stack.add(listNode);
        super.list(element);
        stack.removeLast();
        stack.getLast().appendChild(listNode);
    }

    @Override
    protected void listItem(int level, String bullet, Markup element) {
        Element listItemNode = document.createElement("li");
        applyHTMLStyle(listItemNode, element);
        stack.add(listItemNode);
        doRender(element);
        stack.removeLast();
        stack.getLast().appendChild(listItemNode);
    }

    @Override
    protected void table(Markup element) {
        Element tableElement = document.createElement("table");
        applyHTMLStyle(tableElement, element);
        stack.add(tableElement);
        super.table(element);
        stack.removeLast();
        stack.getLast().appendChild(tableElement);
    }

    @Override
    protected void tableRow(Markup element) {
        Element rowElement = document.createElement("tr");
        applyHTMLStyle(rowElement, element);
        stack.add(rowElement);
        element.getContent().forEach(cell -> {
            Element cellElement = document.createElement(element.getBoolean("header", false) ? "th" : "td");
            stack.add(cellElement);
            doRender(cell);
            stack.removeLast();
            stack.getLast().appendChild(cellElement);
        });
        stack.removeLast();
        stack.getLast().appendChild(rowElement);
    }

    private void applyHTMLStyle(Element html, Markup markup) {
        if (markup.meta().hasNode("html")) {
            MetaUtils.valueStream(markup.meta().getMeta("html")).forEach(pair -> html.setAttribute(pair.getKey(), pair.getValue().stringValue()));
        }
    }
}

File: dataforge-core/src/main/kotlin/hep/dataforge/io/markup/HTMLMarkupRenderer.kt
package hep.dataforge.io.markup

import hep.dataforge.meta.MetaUtils
import org.slf4j.LoggerFactory
import org.w3c.dom.Document
import org.w3c.dom.Element
import java.io.PrintStream
import java.util.*
import javax.xml.parsers.DocumentBuilderFactory
import javax.xml.transform.OutputKeys
import javax.xml.transform.TransformerException
import javax.xml.transform.TransformerFactory
import javax.xml.transform.dom.DOMSource
import javax.xml.transform.stream.StreamResult

/**
 * The html renderer for markup
 * TODO make renderer reusable
 * Created by darksnake on 07-Jan-17.
 */
class HTMLMarkupRenderer(private val stream: PrintStream) : GenericMarkupRenderer() {

    private val document: Document by lazy {
        val factory = DocumentBuilderFactory.newInstance()
        val builder = factory.newDocumentBuilder()
        builder.newDocument()
    }
    private val stack = ArrayDeque<Element>()

    @Throws(TransformerException::class)
    private fun printDocument() {
        val tf = TransformerFactory.newInstance()
        val transformer = tf.newTransformer()
        transformer.setOutputProperty(OutputKeys.OMIT_XML_DECLARATION, "yes")
        transformer.setOutputProperty(OutputKeys.INDENT, "yes")
        transformer.setOutputProperty("{http://xml.apache.org/xslt}indent-amount", "4")
        transformer.setOutputProperty(OutputKeys.ENCODING, "utf8")
        transformer.transform(DOMSource(document), StreamResult(stream))
        stream.flush()
    }

    /**
     * Build DOM without attaching it to document
     *
     * @param document
     * @param markup
     * @return
     */
    private fun buildDOM(markup: Markup): Element {
        stack.clear()
        val root = document.createElement("body")
        stack.add(root)
        doRender(markup)
        stack.removeLast()
        if (stack.size != 0) {
            LoggerFactory.getLogger(javaClass).warn("Node stack not empty after rendering")
        }
        return root
    }

    override fun render(mark: Markup) {
        try {
            document.appendChild(buildDOM(mark))
            printDocument()
        } catch (ex: Exception) {
            throw RuntimeException(ex)
        }

    }

    override fun text(text: String, color: String?, element: Markup) {
        if (color != null) {
            val textElement = document.createElement("font")
            applyHTMLStyle(textElement, element)
            textElement.setAttribute("color", color)
            textElement.textContent = text
            stack.last.appendChild(textElement)
        } else {
            val textElement = document.createTextNode(text)
            stack.last.appendChild(textElement)
        }
    }

    override fun list(element: ListMarkup) {
        val listNode = document.createElement("ul")
        applyHTMLStyle(listNode, element)
        stack.add(listNode)
        super.list(element)
        stack.removeLast()
        stack.last.appendChild(listNode)
    }

    override fun listItem(level: Int, bullet: String, element: Markup) {
        val listItemNode = document.createElement("li")
        applyHTMLStyle(listItemNode, element)
        stack.add(listItemNode)
        doRender(element)
        stack.removeLast()
        stack.last.appendChild(listItemNode)
    }

    override fun table(element: TableMarkup) {
        val tableElement = document.createElement("table")
        applyHTMLStyle(tableElement, element)
        stack.add(tableElement)
        super.table(element)
        stack.removeLast()
        stack.last.appendChild(tableElement)
    }

    override fun tableRow(element: RowMarkup, isHeader: Boolean) {
        val rowElement = document.createElement("tr")
        applyHTMLStyle(rowElement, element)
        stack.add(rowElement)
        element.content.forEach { cell ->
            val cellElement = document.createElement(if (isHeader) "th" else "td")
            stack.add(cellElement)
            doRender(cell)
            stack.removeLast()
            stack.last.appendChild(cellElement)
        }
        stack.removeLast()
        stack.last.appendChild(rowElement)
    }

    private fun applyHTMLStyle(html: Element, markup: Markup) {
        MetaUtils.valueStream(markup.toMeta().getMeta("html")).forEach { pair -> html.setAttribute(pair.first, pair.second.stringValue()) }
    }

    //    @Override
    //    protected void header(Markup element) {
    //        Element headerElement = document.createElement("h" + element.getInt("level", 1));
    //        headerElement.setTextContent(element.);
    //        applyHTMLStyle(headerElement, element);
    //        stack.getLast().appendChild(headerElement);
    //    }
}
